src/delphi/train/utils.py

```diff
@@ -0,0 +1,295 @@
+    return model
+
+
+def config_to_model(config: ModelConfig) -> PreTrainedModel:
+    # get ModelType object from name ('llama' -> ModelType(...))
+    delphi_config = get_delphi_config(config)
+    model_type = ModelTypes.get(config.model_type)
+    return model_type.model(model_type.config(**asdict(delphi_config)))
+
+
+def initialize_model_training_state(
+    config: GigaConfig, device: torch.device
+) -> ModelTrainingState:
+    iter_num = 0
+    local_iter_num = 0
+    best_val_loss = 1e9
+    running_mfu = -1.0
+    t0 = time.time()
+    if config.init_from == "scratch":
+        # init a new model from scratch
+        print("Initializing a new model from scratch")
+        model = config_to_model(config.model_config)
+        checkpoint = None
+    # TODO: resume from huggingface model
+    elif config.init_from == "resume":
+        print(f"Resuming training from {config.output_dir}")
+        checkpoint = torch.load(
+            Path(config.output_dir) / "ckpt.pt", map_location=device
+        )
+        model = load_model(config, checkpoint)
+        iter_num = checkpoint["iter_num"]
+        best_val_loss = checkpoint["best_val_loss"]
+    model.to(device)  # type: ignore
+    # optimizer
+    optimizer = get_optimizer(
+        model=model,
+        config=config,
+        checkpoint=checkpoint
+        if checkpoint is not None and "optimizer" in checkpoint  # type: ignore
+        else None,
+    )
+    epoch = checkpoint.get("epoch", 0) if checkpoint is not None else 0
+    step = checkpoint.get("step", 0) if checkpoint is not None else 0
+    checkpoint = None  # free up memory
+    return ModelTrainingState(
+        model=model,
+        optimizer=optimizer,
+        iter_num=iter_num,
+        local_iter_num=local_iter_num,
+        best_val_loss=best_val_loss,
+        running_mfu=running_mfu,
+        t0=t0,
+        epoch=epoch,
+        step=step,
+    )
+
+
+def load_delphi_training_dataset(split: str, limit: int = -1):
+    """For training, we want (X, Y) pairs, where X is a chunk of text and Y is the next token.)
+    To construct this, we take the original tokenized dataset, break it into max_seq_len+1 length chunks,
+    and then take [:-1] as X and [1:] as Y.
+    """
+    if limit == -1:
+        ds = load_delphi_dataset(constants.TOKENIZED_CORPUS_DATASET, split)
+    else:
+        ds = load_delphi_dataset(constants.TOKENIZED_CORPUS_DATASET, split).select(
+            range(limit)
+        )
+    ds.set_format("torch")
+    return ds
+
+
+def get_next_xy(
+    train_batch_iter: Generator,
+    device: torch.device
+    # train_batch_iter: Generator[dict[str, list[int]], None, None], device: torch.device
+) -> tuple[torch.Tensor, torch.Tensor]:
+    data = next(train_batch_iter).to(device)
+    # X and Y NEED to be contigious. llama2c's implementation involves
+    # calling .view on them, which breaks if they're not contigious
+    X, Y = data[:, :-1].contiguous(), data[:, 1:].contiguous()
+    return X, Y
```

Comment:
Given the next batch of samples from the dataset, break it into X and Y (label) tensors for training/validation

Reply:
we're not using llama2c anymore, do they still need to be contiguous()?
also, can we have lowercase x, y? 😅

Reply:
lol, good point. fixed